### PR TITLE
Add docs for desired balance allocator

### PR DESCRIPTION
**Pull request by DaveCTurner**
These docs cover the new allocator and the settings controlling the heuristics for combining disk usage and write load into the overall weight.

### CODE REVIEW

File: docs/reference/modules/cluster/shards_allocation.asciidoc [hunk 1]
```diff
@@ -98,13 +98,33 @@ Specify when shard rebalancing is allowed:
...
+
+`cluster.routing.allocation.type`::
++
+--
+Selects the algorithm used for computing the cluster balance. Defaults to
+`desired_balance` which selects the _desired balance allocator_. This allocator
+runs a background task which computes the desired balance of shards in the
+cluster by repeatedly simulating shard movements until no shard movement
+improves the balance of the cluster. Once this background task completes, {es}
+moves shards to their desired locations.
```
Comment (DaveCTurner) [inline docs/reference/modules/cluster/shards_allocation.asciidoc:117]: This is obviously pretty brief but I'm not sure we need to go into much more detail here today. There's not much user-visible stuff in this area yet. I expect we'll extend this and perhaps break it out into its own section once we have more APIs and things.

Comment (idegtiarenko) [reply docs/reference/modules/cluster/shards_allocation.asciidoc:117]:
```
computes the desired balance of shards in the
cluster by repeatedly simulating shard movements until no shard movement
improves the balance of the cluster
```
this sounds a lot like implementation detail to me. Should we consider:
```
computes the desired balance of shards in the cluster until shard counts and their 
resource usages are as close to even distribution as allowed by allocation settings
```
?

File: docs/reference/modules/cluster/shards_allocation.asciidoc [hunk 2]
```diff
@@ -57,17 +57,18 @@ one of the active allocation ids in the cluster state.
...
-in your cluster to improve its balance. Rebalancing obeys all other shard
-allocation rules such as <<cluster-shard-allocation-filtering,allocation
-filtering>> and <<forced-awareness,forced awareness>> which may prevent it from
-completely balancing the cluster. In that case, rebalancing strives to achieve
-the most balanced cluster possible within the rules you have configured. If you
-are using <<data-tiers,data tiers>> then {es} automatically applies allocation
-filtering rules to place each shard within the appropriate tier. These rules
-mean that the balancer works independently within each tier.
+A cluster is _balanced_ when it has an equal number of shards on each node, with
+all nodes needing equal resources, without having a concentration of shards from
```
Comment (idegtiarenko) [inline docs/reference/modules/cluster/shards_allocation.asciidoc:61]: Should we give an example for resources to be more specific here (eg disk usage and may be ingest load)?

File: docs/reference/modules/cluster/shards_allocation.asciidoc [hunk 3]
```diff
@@ -114,28 +134,46 @@ of shards, and then moving shards between nodes to reduce the weight of the
...
-     (<<dynamic-cluster-setting,Dynamic>>)
-     Defines the weight factor for the number of shards per index allocated
-      on a specific node (float). Defaults to `0.55f`. Raising this raises the
-      tendency to equalize the number of shards per index across all nodes in
-      the cluster.
+(float, <<dynamic-cluster-setting,Dynamic>>)
+Defines the weight factor for the number of shards per index allocated to each
+node. Defaults to `0.55f`. Raising this value increases the tendency of {es} to
+equalize the number of shards of each index across nodes ahead of the other
+balancing variables.
```
Comment (idegtiarenko) [inline docs/reference/modules/cluster/shards_allocation.asciidoc:155]: May be `Raising this value increases the tendency of {es} to spread shards of the same index to a different nodes ahead of the other
balancing variables`?

**Review by idegtiarenko — APPROVED** [review]
Left couple suggestions, otherwise, :+1:

File: docs/reference/modules/cluster/shards_allocation.asciidoc (hunk 1 above)
Comment (henningandersen) [inline docs/reference/modules/cluster/shards_allocation.asciidoc:117]:
```suggestion
`desired_balance` which selects the _desired balance allocator_. This allocator
runs a background task which computes the desired balance of shards in the
cluster. Once this background task completes, {es}
moves shards to their desired locations.
```

Comment (henningandersen) [reply docs/reference/modules/cluster/shards_allocation.asciidoc:117]: Agree, I suggested an even simpler version of it.

File: docs/reference/modules/cluster/shards_allocation.asciidoc [hunk 4]
```diff
@@ -114,28 +134,46 @@ of shards, and then moving shards between nodes to reduce the weight of the
...
+`cluster.routing.allocation.balance.disk_usage`::
+(float, <<dynamic-cluster-setting,Dynamic>>)
+Defines the weight factor for balancing shards according to their predicted disk
+size in bytes. Defaults to `2e-11f`. Raising this value increases the tendency
+of {es} to equalize the total disk usage across nodes ahead of the other
+balancing variables.
+
+`cluster.routing.allocation.balance.write_load`::
+(float, <<dynamic-cluster-setting,Dynamic>>)
+Defines the weight factor for the write load of each shard, in terms of the
```
Comment (henningandersen) [inline docs/reference/modules/cluster/shards_allocation.asciidoc:166]: Is this the place where we should mention that only the write-index of data streams have a write-load?

**Review by henningandersen — APPROVED** [review]
LGTM.

File: docs/reference/modules/cluster/shards_allocation.asciidoc (hunk 1 above)
Comment (DaveCTurner) [reply docs/reference/modules/cluster/shards_allocation.asciidoc]: Makes sense, thanks.

File: docs/reference/modules/cluster/shards_allocation.asciidoc (hunk 4 above)
Comment (DaveCTurner) [reply docs/reference/modules/cluster/shards_allocation.asciidoc:174]: I tried but didn't really like it. I added a paragraph further up about how resource usage is estimated.

File: docs/reference/modules/cluster/shards_allocation.asciidoc (hunk 3 above)
Comment (DaveCTurner) [reply docs/reference/modules/cluster/shards_allocation.asciidoc:163]: I also prefer `spread` (or `spread out`) over `equalize` but it doesn't really fit with the `ahead of the other balancing variables` so well and I'm struggling to find an alternative wording.

File: docs/reference/modules/cluster/shards_allocation.asciidoc (hunk 2 above)
Comment (DaveCTurner) [reply docs/reference/modules/cluster/shards_allocation.asciidoc:61]: I'm ok with being vague here - we go into more detail lower down, and the general principle really applies to all resources even if we don't consider most of them today.

**Pull request merged**